### PR TITLE
Allow passing a message to ClientCall.cancel().

### LINKED ISSUE #1221: Consider allowing clients to provide cancellation details
When designing `Context` we decided that `cancel(Throwable)` could be useful for debugging where a cancellation came from. In general, I'm not a fan of using that information for changing behavior, but it is very helpful for debugging.

Currently `ClientCall.cancel()` does not accept any arguments, because there is no way for the client to inform the server of any reasoning (and aren't too interested in that scenario, because servers can't really trust clients too much). When reviewing #1203, I saw that `Context` auto-propagation can make it quite opaque why a call was cancelled. We could have `ClientCall.cancel(String desc, Throwable t)` which will end up notifying the application `onClose()` with a `Status.CANCELLED.withDescription(desc).withCause(t)`.

That would allow us to propagate a `Context`'s failure exception as well.

### PR DESCRIPTION
**Pull request by zhangkun83**
Alternative to #1714
Resolves #1221, based on its latest comment.

### CODE REVIEW

**Comment by zhangkun83** [comment]
@carl-mastrangelo @louiscryan FYI

File: core/src/main/java/io/grpc/ForwardingClientCall.java [hunk 1]
```diff
@@ -51,8 +51,8 @@ public void request(int numMessages) {
   }
 
   @Override
-  public void cancel() {
-    delegate().cancel();
+  public void cancel(String message) {
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ForwardingClientCall.java]: You'll still need to forward regular cancel.

File: stub/src/test/java/io/grpc/stub/ClientCallsTest.java [hunk 2]
```diff
@@ -101,7 +102,7 @@
     verify(call).start(listenerCaptor.capture(), any(Metadata.class));
     ClientCall.Listener<String> listener = listenerCaptor.getValue();
     future.cancel(true);
-    verify(call).cancel();
+    verify(call).cancel(eq("GrpcFuture was cancelled"));
```
Comment (carl-mastrangelo) [inline stub/src/test/java/io/grpc/stub/ClientCallsTest.java]: no need to use eq.  It should work by default.

**Comment by zhangkun83** [comment]
#1714 has been merged. Abandoning this one.

**Pull request closed without merging**